### PR TITLE
[mono] show whole trace of nested exceptions

### PR DESCRIPTION
**Pull request by paulloz**
Issue #15886

The debugger now shows the whole trace of nested exception (it was only showing the top level for now).
capture du 2018-02-24 17-02-48

I'm not sure 0c63b15 should be included as it feels a bit hacky, is not really mandatory and could feel like module specific code outside of the module itself. Feel free to tell me to delete it.

### CODE REVIEW

File: editor/script_editor_debugger.cpp [hunk 1]
```diff
@@ -1646,7 +1646,7 @@ void ScriptEditorDebugger::_error_selected(int p_idx) {
 		md.push_back(st[i + 1]);
 		md.push_back(st[i + 2]);
 
-		String str = func + " in " + script.get_file() + ":line " + itos(line);
+		String str = func + (line != NULL ? (" in " + script.get_file() + ":line " + itos(line)) : "");
```
Comment (neikeq) [inline editor/script_editor_debugger.cpp]: This is comparing line to 0. The script debugger handles lines as beginning with 1, so using 0 is fine here. I would use the script path for this though. Something like this:

```
String str = func;
if (script.length()) {
    str += " in " + script.get_file();
    if (line != 0) // do not show _":line 0"_ if the line if it's unknown
        str += ":line " + itos(line);
}
```

Comment (paulloz) [reply editor/script_editor_debugger.cpp]: I changed this, and took the opportunity to do the same thing with the tooltip.

File: modules/mono/mono_gd/gd_mono_utils.cpp [hunk 2]
```diff
@@ -420,37 +420,56 @@ void print_unhandled_exception(MonoObject *p_exc, bool p_recursion_caution) {
 	if (!ScriptDebugger::get_singleton())
 		return;
 
-	GDMonoClass *st_klass = CACHED_CLASS(System_Diagnostics_StackTrace);
-	MonoObject *stack_trace = mono_object_new(mono_domain_get(), st_klass->get_mono_ptr());
+	ScriptLanguage::StackInfo separator;
+	separator.file = "";
+	separator.func = "--- " + TTR("End of inner exception stack trace") + " ---";
+	separator.line = NULL;
```
Comment (neikeq) [inline modules/mono/mono_gd/gd_mono_utils.cpp]: This should be 0 to be clear.

Comment (paulloz) [reply modules/mono/mono_gd/gd_mono_utils.cpp]: Sorry, didn't think of that.

**Comment by neikeq** [comment]
Thanks! :)

**Pull request merged**